File: core/Lumino/Controls/ScrollViewerWithTouch.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Controls.Primitives;
using System.Windows.Input;
using System.Windows.Media;

namespace Lumino.Controls
{
    /// <summary>
    /// ScrollViewer를 터치 인터페이스에 자연스럽게 대응할 수 있도록 개선합니다.
    /// 출처 : http://goo.gl/6yatQy
    /// </summary>
    class ScrollViewerWithTouch : ScrollViewer
    {
        /// <summary>
        /// Original panning mode.
        /// </summary>
        private PanningMode panningMode;

        /// <summary>
        /// Set panning mode only once.
        /// </summary>
        private bool panningModeSet;

        /// <summary>
        /// Initializes static members of the <see cref="ScrollViewerWithTouch"/> class.
        /// </summary>
        static ScrollViewerWithTouch()
        {
            DefaultStyleKeyProperty.OverrideMetadata(typeof(ScrollViewerWithTouch), new FrameworkPropertyMetadata(typeof(ScrollViewerWithTouch)));
        }

        protected override void OnManipulationCompleted(ManipulationCompletedEventArgs e)
        {
            base.OnManipulationCompleted(e);

            // set it back
            this.PanningMode = this.panningMode;
        }

        protected override void OnManipulationStarted(ManipulationStartedEventArgs e)
        {
            // figure out what has the user touched
            var result = VisualTreeHelper.HitTest(this, e.ManipulationOrigin);
            if (result != null && result.VisualHit != null)
            {
                var hasButtonParent = this.HasButtonParent(result.VisualHit);

                // if user touched a button then turn off panning mode, let style bubble down, in other case let it scroll
                this.PanningMode = hasButtonParent ? PanningMode.None : this.panningMode;
            }

            base.OnManipulationStarted(e);
        }

        protected override void OnTouchDown(TouchEventArgs e)
        {
            // store panning mode or set it back to it's original state. OnManipulationCompleted does not do it every time, so we need to set it once more.
            if (this.panningModeSet == false)
            {
                this.panningMode = this.PanningMode;
                this.panningModeSet = true;
            }
            else
            {
                this.PanningMode = this.panningMode;
            }

            base.OnTouchDown(e);
        }

        private bool HasButtonParent(DependencyObject obj)
        {
            var parent = VisualTreeHelper.GetParent(obj);

            if ((parent != null) && (parent is ButtonBase) == false)
            {
                return HasButtonParent(parent);
            }

            return parent != null;
        }
    }
}

File: widgets/music/Music/Controls/ScrollViewerWithTouch.vb

Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports System.Windows
Imports System.Windows.Controls
Imports System.Windows.Controls.Primitives
Imports System.Windows.Input
Imports System.Windows.Media

Class ScrollViewerWithTouch
    Inherits ScrollViewer
    ''' <summary>
    ''' Original panning mode.
    ''' </summary>
    Private _PanningMode As PanningMode

    ''' <summary>
    ''' Set panning mode only once.
    ''' </summary>
    Private _PanningModeSet As Boolean

    ''' <summary>
    ''' Initializes static members of the <see cref="ScrollViewerWithTouch"/> class.
    ''' </summary>
    Shared Sub New()
        DefaultStyleKeyProperty.OverrideMetadata(GetType(ScrollViewerWithTouch), New FrameworkPropertyMetadata(GetType(ScrollViewerWithTouch)))
    End Sub

    Protected Overrides Sub OnManipulationCompleted(e As ManipulationCompletedEventArgs)
        MyBase.OnManipulationCompleted(e)

        ' set it back
        PanningMode = _PanningMode
    End Sub

    Protected Overrides Sub OnManipulationStarted(e As ManipulationStartedEventArgs)
        ' figure out what has the user touched
        Dim result = VisualTreeHelper.HitTest(Me, e.ManipulationOrigin)
        If result IsNot Nothing AndAlso result.VisualHit IsNot Nothing Then
            Dim hasButtonParent = Me.HasButtonParent(result.VisualHit)

            ' if user touched a button then turn off panning mode, let style bubble down, in other case let it scroll
            PanningMode = If(hasButtonParent, PanningMode.None, _PanningMode)
        End If

        MyBase.OnManipulationStarted(e)
    End Sub

    Protected Overrides Sub OnTouchDown(e As TouchEventArgs)
        ' store panning mode or set it back to it's original state. OnManipulationCompleted does not do it every time, so we need to set it once more.
        If _PanningModeSet = False Then
            _PanningMode = PanningMode
            _PanningModeSet = True
        Else
            PanningMode = _PanningMode
        End If

        MyBase.OnTouchDown(e)
    End Sub

    Private Function HasButtonParent(obj As DependencyObject) As Boolean
        Dim parent = VisualTreeHelper.GetParent(obj)

        If (parent IsNot Nothing) AndAlso (TypeOf parent Is ButtonBase) = False Then
            Return HasButtonParent(parent)
        End If

        Return parent IsNot Nothing
    End Function
End Class